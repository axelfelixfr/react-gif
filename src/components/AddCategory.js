import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({setCategories}) => {

    // Le damos un valor por defecto al input con el useState
    const [inputValue, setInputValue] = useState(''); // Mandamos como estado por defecto un string vacio

    // Realizamos una función para el input
    const handleInputChange = (e) => {
        // Hacemos uso del método setInputValue del useState para cambiar su value
        setInputValue(e.target.value);
    };

    const handleSubmit = (e) => {
        e.preventDefault(); // Quitamos el evento por default
        // Verificamos que por lo menos haya un string con longitud de tres para no mandar datos vacíos
        if(inputValue.trim().length > 2){
            // Accedemos a las categorías y le agregamos lo que esta en inputValue
            setCategories(cats => [inputValue, ...cats]);
            // Borramos el contenido del input
            setInputValue('');
        }
    };

    return (
        // Le pasamos el evento handleSubmit al formulario
        <form onSubmit={handleSubmit}>
            <input 
                type="text"
                // Le pasamos el value del useState
                value={ inputValue }
                // Le pasamos la función que se ejecutara cada que cambie el input
                onChange={ handleInputChange }
            />
        </form>
    )
}

// setCategories pasa a ser requerida
AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}