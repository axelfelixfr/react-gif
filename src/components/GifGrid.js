import React from 'react';
import PropTypes from 'prop-types';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {
    // A la data la renombramos images
    const { data:images, loading } = useFetchGifs(category);

    return (
        <>
            {/* Le pasamos el nombre de la categoría (de la búsqueda que se hizo) */}
            <h3 className="animate__animated animate__fadeIn">{category}</h3>
            
            {/* El loading debe estar en true para que aparezca el parrafo */}
            { loading && <p className="animate__animated animate__flash">Loading</p> }
            <div className="card-grid">
            
            {/* Extraemos todos los gifs con map, y le pasamos al componente <GifGridItem /> la data de cada gif que se recoge */}
                {
                    images.map(img => (
                        <GifGridItem 
                            key={ img.id }
                            { ...img }
                        />
                    ))              
                }
            
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}