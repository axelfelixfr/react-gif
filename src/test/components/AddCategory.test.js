import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import { AddCategory } from '../../components/AddCategory';

describe('Pruebas en el componente <AddCategory />', () => {

    // Con el objeto jest podemos simular una función con el método fn()
    const setCategories = jest.fn();
    let wrapper = shallow(<AddCategory setCategories={setCategories} />);

    // Antes de cada prueba va a realizar esta funci´ón
    beforeEach(() => {
        // Limpiamos las acciones realizadas antes de cada prueba con clearAllMocks
        jest.clearAllMocks();
        // Inicializamos nuevamente el wrapper por cada prueba
        wrapper = shallow(<AddCategory setCategories={setCategories} />);
    });

    test('Debe de mostrarse correctamente el componente', () => {
        // Comparamos que el componente haga match con el snapshot 
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de cambiar la caja de texto', () => {
        // Seleccionamos el input con el método find()
        const input = wrapper.find('input');
        const value = 'Hola mundo!!';

        // Simulamos el evento change cambiando el value del input
        input.simulate('change', {target: { value }});

        // Esperamos que el parrafo tenga el mismo contenido (value) que el input
        expect(wrapper.find('p').text().trim()).toBe(value);
    });

    test('NO debe de mandar la información con submit', () => {
        // Simulamos el evento submit (enviar el formulario) pero con el input vacio
        wrapper.find('form').simulate('submit', { preventDefault(){} });

        // La función setCategories nunca debio ser llamada así que la negamos con 'not'
        expect(setCategories).not.toHaveBeenCalled();
    });

    test('Debe de llamar el setCategories y limpiar la caja de texto', () => {
        const value = 'Hola mundo!!';

        // Le mandamos al input el value de 'Hola mundo!!!'
        wrapper.find('input').simulate('change', {target: {value}});

        // Simulamos enviar el formulario con el evento submit
        wrapper.find('form').simulate('submit', { preventDefault(){} });

        // Esperamos que la función setCategories haya sido llamada
        expect(setCategories).toHaveBeenCalled();

        // Esperamos que haya sido llamada una vez
        expect(setCategories).toHaveBeenCalledTimes(1);

        // Esperamos que se llamada una función y no cualquier otra cosa
        expect(setCategories).toHaveBeenCalledWith(expect.any(Function));

        // Esperamos que el value del input este vacío nuevamente
        expect(wrapper.find('input').prop('value')).toBe('');
    });
});