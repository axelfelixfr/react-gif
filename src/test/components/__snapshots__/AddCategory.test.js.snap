// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <AddCategory /> Debe de mostrarse correctamente el componente 1`] = `
<form
  onSubmit={[Function]}
>
  <p
    className="displayPrueba"
  />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
